src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs

```diff
@@ -1309,85 +1311,169 @@ public static unsafe int LastIndexOfAny(ref byte searchSpace, byte value0, byte
 
         // Optimized byte-based SequenceEquals. The "length" parameter for this one is declared a nuint rather than int as we also use it for types other than byte
         // where the length can exceed 2Gb once scaled by sizeof(T).
-        [MethodImpl(MethodImplOptions.AggressiveOptimization)]
-        public static unsafe bool SequenceEqual(ref byte first, ref byte second, nuint length)
+        public static bool SequenceEqual(ref byte first, ref byte second, nuint length)
         {
-            IntPtr offset = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr lengthToExamine = (IntPtr)(void*)length;
-
-            if ((byte*)lengthToExamine >= (byte*)sizeof(UIntPtr))
+            // Use nint for arithmetic to avoid unnecessary 64->32->64 truncations
+            if (length >= sizeof(nuint))
             {
-                // Only check that the ref is the same if buffers are large, and hence
-                // its worth avoiding doing unnecessary comparisons
-                if (Unsafe.AreSame(ref first, ref second))
-                    goto Equal;
+                // Conditional jmp foward to favor shorter lengths. (See comment at "Equal:" label)
+                // The longer lengths can make back the time due to branch misprediction
+                // better than shorter lengths.
+                goto Longer;
+            }
 
-                if (Vector.IsHardwareAccelerated && (byte*)lengthToExamine >= (byte*)Vector<byte>.Count)
+#if TARGET_64BIT
+            // On 32-bit, this will never be true since sizeof(nuint) == 4
+            if (length >= sizeof(uint))
+            {
+                nint offset = 0;
+                if (length > sizeof(uint))
                 {
-                    lengthToExamine -= Vector<byte>.Count;
-                    while ((byte*)lengthToExamine > (byte*)offset)
+                    // Set offset for next compare to sizeof(uint) from end
+                    offset = (nint)length - sizeof(uint);
+                    // Compare start
+                    if (LoadUInt(ref first) != LoadUInt(ref second))
                     {
-                        if (LoadVector(ref first, offset) != LoadVector(ref second, offset))
-                        {
-                            goto NotEqual;
-                        }
-                        offset += Vector<byte>.Count;
+                        goto NotEqual;
                     }
-                    return LoadVector(ref first, lengthToExamine) == LoadVector(ref second, lengthToExamine);
                 }
 
-                Debug.Assert((byte*)lengthToExamine >= (byte*)sizeof(UIntPtr));
-
-                lengthToExamine -= sizeof(UIntPtr);
-                while ((byte*)lengthToExamine > (byte*)offset)
+                // Compare end
+                return LoadUInt(ref first, offset) == LoadUInt(ref second, offset);
+            }
+#endif
+            if (length >= sizeof(ushort))
```

Comment:
For these last two blocks, you may want to consider the below to help reduce codegen size and the total number of jumps required. I haven't tested it yet so I don't know if it will improve throughput.

```cs
{
    nint offset = 0;
    uint differentBits = 0;
    if ((length & 2) != 0)
    {
        offset = 2;
        differentBits = LoadUShort(ref first);
        differentBits -= LoadUShort(ref second);
    }
    if ((length & 1) != 0)
    {
        differentBits |= (uint)Unsafe.AddByteOffset(ref first, offset) - (uint)Unsafe.AddByteOffset(ref second, offset);
    }
    return (differentBits == 0);
}
```

In theory this should reduce the overall codegen size and number of jumps when the total length is between 0 .. 3 bytes. Pseudo-asm:

```asm
xor offset, offset
xor differentBits, differentBits
test length, 2
jz DoByteCheck
add offset, 2
movzx differentBits, word ptr [first]
movzx tmp, word ptr [second]
sub differentBits, tmp
DoByteCheck:
test length, 1
jz AfterByteCheck
movzx tmp, byte ptr [first + offset]
movzx tmp2, byte ptr [second + offset]
sub tmp, tmp2
or differentBits, tmp
AfterByteCheck:
test differentBits, differentBits
setnz al
movzx eax, al
ret
```